### PR TITLE
DPL:  event based processing

### PR DESCRIPTION
**Pull request by ktf**
This is a major refactoring of the DPL driver and DataProcessingDevice, integrating libuv as a backend to do event driven I/O rather than the old busy loop. In particular this provides:

* DataProcessingDevice does not need to busy loop anymore. When ZeroMQ activity is detected on the socket by the libuv poller, we enter the standard event processing of DPL, processing all pending messages. This should ensure that we busy loop as fast as possible when data needs to be processed and we block on the epoll / kqueue call when nothing is happening. @Barthelemy @davidrohr @knopers8 this should remove completely our issue with "what delay to use to ensure best performance without consuming too much resources".

* The driver itself is fully event driven by libuv, again making it more responsive and avoiding unnecessary busy looping. In particular logs from the devices are processed only when available.

* @davidrohr this should allow you to put tasks on the libuv loop and run them multithreaded. We need to think of a way to make sure that the continuations for those tasks do the send message correctly.

### CODE REVIEW

**Comment by ktf** [comment]
@knopers8 could you run your performance benchmark on this to see if there is any major regression?

**Comment by knopers8** [comment]
> @knopers8 could you run your performance benchmark on this to see if there is any major regression?

Sure, that will be interesting to see. Though I have to update it to new metrics format first...

Does this refactoring take Timers into account? Sorry, I don't have time to have a closer look myself.

File: Framework/Core/src/DataProcessingDevice.cxx [hunk 1]
```diff
@@ -236,17 +267,107 @@ void DataProcessingDevice::InitTask()
       pendingRegionInfos.push_back(info);
     });
   }
+
+  // Add a signal manager for SIGUSR1, SIGUSR2, so that
```
Comment (davidrohr) [inline Framework/Core/src/DataProcessingDevice.cxx:271]: So that what?

**Comment by ktf** [comment]
@knopers8 in principle yes, albeit still in a synchronous way. We can have proper libuv timers at some point as well.

**Comment by knopers8** [comment]
@ktf What do you mean by synchronous? Synchronous to incoming data?

**Comment by ktf** [comment]
...or to the ability to write on a socket. Basically now we watch for two events:

* Incoming data on the input channel socket
* Ability to write on the outgoing socket

The first one is data driven and we keep reading data as long as there was something happening in the previous iteration.

The second one gets periodically trigger by libuv whenever the output socket can be written to. In both cases what we do is to "unblock" the event loop which then has the same "synchronous" steps as before.

Basically at this point we do not use libuv to actually act on the event, but merely as a way to step through the old "busy loop".

File: Framework/Core/src/DataProcessingDevice.cxx (hunk 1 above)
Comment (ktf) [reply Framework/Core/src/DataProcessingDevice.cxx]: fixed...

**Comment by ktf** [comment]
Ok, on my side I think this is ready to be merged. Further improvements / added features can come in a later PR.

@knopers8 @davidrohr could you check there is no regression in performance on your side?

@knopers8 is the testWorkflow supposed to work?

**Comment by davidrohr** [comment]
Didn't test it yet, but I can run some triggered pp events. They were triggering the performance regression with the exponential back-off. No idea what else to test. I can test the full reco workflow, just to verify it still works.

**Comment by ktf** [comment]
@davidrohr that would be great.

**Comment by davidrohr** [comment]
@ktf : Performance with small events is fine.
I tested full TPC + ITS workflow, couldn't run TOF since I have to simulate new data with RDH6 I assume. Right now the compressor is failing with my old dataset.
But your PR seems to be OK for me.

**Comment by shahor02** [comment]
I don't see any difference in running largest w/flow with 9 TFs.

**Comment by ktf** [comment]
@shahor02 thanks. Do you know if the TPCFastTransform test has issues? It seems unrelated to this PR...

**Comment by shahor02** [comment]
Yes, I have seen it failing randomly in different PRs

**Comment by ktf** [comment]
@MichaelLettrich looks like the macOS builder needs a bump...

**Comment by knopers8** [comment]
> @knopers8 @davidrohr could you check there is no regression in performance on your side?

Yes, I will try find some time today for that.

> @knopers8 is the testWorkflow supposed to work?

In principle yes. I think we have set a too rigorous timeout which might not be enough on an overloaded CI machine. Anyway, we have increased it, but we have not released a new version yet.
However, this test uses Timers, so it potentially might be broken because of PR. I will check it manually as well.

**Comment by ktf** [comment]
Ok, thanks.

**Comment by knopers8** [comment]
I see around 5% drop in maximum message rates, but it could be just statistics. I had a look at Disptacher with `perf`, it is mostly locks and mutexes (mutices?), including this one which is not that high without this PR:
```
boost::intrusive::multiset_impl<boost::intrusive::bhtraits<boost::interprocess::rbtree_best_fit<boost::interprocess::mutex_family, boost::interprocess::offset_ptr<void, long, unsigned long, 0ul>
```
image
 

QC is fine, including Timers. The test that failed in CI works locally.

**Comment by ktf** [comment]
@knopers8 thanks! My conclusion from your analysis is that we can merge.

The mutex comes from shared memory transport which is in principle untouched by this PR, so it's either within the statistical noise (which would be my best guess) or this PR might slightly be more efficient in making a consumer pick up new data, which might increase the contention on the mutex. This would effectively resulting in lower performance because concurrent access to a mutex is way more expensive than separate unblocked access. The latter is of course just speculation.

**Comment by ktf** [comment]
BTW, it should have fixed the high CPU usage when a device is idle, could you confirm?

**Comment by Barthelemy** [comment]
@ktf I have to recompile to test it, so it will take some time

**Comment by knopers8** [comment]
With `o2-qc-run-basic` I see that Dispatcher and QC task use 100%. They both have Timers so my guess is that they are still constantly checking if certain time has passed. Anyway, these are top 15 offenders in such case:
image

**Comment by ktf** [comment]
Can you provide me a simple example for the dispatcher?

**Comment by knopers8** [comment]
`o2-testworkflows-datasampling-pod-and-root` should work out of the box. Just tried it, I see 100% even for small message rates.

**Comment by ktf** [comment]
Ok. Looking into it. I think I know the problem...
